SPEC.md

```diff
@@ -0,0 +1,150 @@
+# CNI Specification
+Note that the prime source of information about CNI is to be considered the reference implementation and test suite.
+This document is a reading companion (rather than the source of truth) that should help expediate the development of implementations.
+
+## Core Language
+A CNI document consists of an arbitrary amount of statements.
+A statement may either be a section declaration, key-value pair, or key-rawvalue pair.
+
+Note that it is valid to have multiple statements assigning to the same key.
+In the event this happens, the last key definition wins.
+For example,
+```
+sub.source = main.zip
+[sub]
+source = src.zip
+```
+will result in the `sub.source` key having the value of `src.zip`.
+
+### Key
+A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
+Keys may contain dots, but may not start or end with them - dots semantically signify sections.
+Two dots in a row are disallowed (due to being nonsensical, it would imply a section with no name), but implementations do not need to explicitly disallow them.
+
+### Section Declaration
+A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
```

Comment:
I think to be consistent we should allow horizontal whitespace instead of just spaces.
```suggestion
It's a key, an equals, and a value, with arbitrary amounts of horizontal whitespace before and after the `=`.
```

Reply:
I think core should only require people to check for ' ' as spacings, to simplify some parsers.
So if anything I would rather reverse it (if they do non-significant whitespace properly they end up auto-implementing the flexspace and tabulation extensions, eh?)
OTOH, users might want to be able to insert hard tabs.
I'll have to think about this.
Might end up removing tabulation/flexspace and adding those into core tbh.